File: adventjs/2025/09/solution.ts
type Result = "fail" | "crash" | "success"

function moveReno(board: string, moves: string): Result {
  const grid = board
    .split("\n")
    .slice(1, -1)
    .map((l) => l.trim())
    .filter(Boolean)
  if (!grid.length) return "fail"

  const dirs: Record<string, number[]> = {
    L: [0, -1],
    R: [0, 1],
    U: [-1, 0],
    D: [1, 0],
  }
  const start = grid.findIndex((row) => row.includes("@"))
  if (start < 0) return "fail"

  type State = { r: number; c: number; picked: boolean; done: Result | null }

  const result = [...moves].reduce<State>(
    (s, m) => {
      if (s.done) return s
      const d = dirs[m] ?? [0, 0]
      const r = s.r + d[0],
        c = s.c + d[1]
      const cell = grid[r]?.[c]
      if (!cell || cell === "#")
        return { ...s, done: s.picked ? "success" : "crash" }
      return { r, c, picked: s.picked || cell === "*", done: null }
    },
    { r: start, c: grid[start].indexOf("@"), picked: false, done: null }
  )

  return result.done ?? (result.picked ? "success" : "fail")
}

File: adventjs/2025/09/solution.py
from typing import List, Literal

def move_reno(board: str, moves: str) -> Literal['fail', 'crash', 'success']:
  grid = [l.strip() for l in board.split('\n')[1:-1] if l.strip()]
  if not grid:
    return 'fail'
  
  dirs = {'L': (0, -1), 'R': (0, 1), 'U': (-1, 0), 'D': (1, 0)}
  row = next((i for i, r in enumerate(grid) if '@' in r), -1)
  if row == -1:
    return 'fail'
  col = grid[row].index('@')
  picked_up = False
  
  for move in moves:
    dr, dc = dirs.get(move, (0, 0))
    row += dr
    col += dc
    
    in_bounds = 0 <= row < len(grid) and 0 <= col < len(grid[0])
    if not in_bounds or grid[row][col] == '#':
      return 'success' if picked_up else 'crash'
    
    if grid[row][col] == '*':
      picked_up = True
  
  return 'success' if picked_up else 'fail'
